### PR TITLE
Fixed #10038 (Tornado, UI/UX, Search feature improvement)

### LINKED ISSUE #10038: [DOCS] Page wise display of documentation search
**Is your feature request related to a problem?**
Yes. I searched for a relatively simple query in the documentation search bar of https://docs.bokeh.org, and it took very long to load the results. In my second try, the results weren't even loading, I'm afraid. These are displayed in an unordered list which fills the entire page up. It might get frustrating to read through everything to find the answer to the input query.   

**Describe the solution you'd like**
I would suggest displaying the fetched results in a page wise format, the way most search engines do it. Relevance weighted sorted answer, shown page wise. Fill up only the current page of about 20 to 30 odd query results, and depending on whether the user wants to see the other pages, load them.

**Describe alternatives you've considered**
If not a page wise result, a folder wise result would also benefit, which leaves the option to the user to navigate where he/she wants to. A custom google search may also help.

**Additional context**
issue

### PR DESCRIPTION
**Pull request by nimishbongale**
addresses #10038

1. Makes it compatible with Python 3.8 + Windows 10 config. (complete)

There are more changes to the `sphinx/_build/search.html`, `sphinx/_build/index.html` and `latest/_static/searchtools.js`, but since these generate only once a `sphinx-build` command is run, I am not sure how to go about contributing to those.    

Do let me know how that is supposed to be done! Maybe I can commit the code in the `docs.bokeh.org` repository?

Thanks

### CODE REVIEW

File: sphinx/docserver.py [hunk 1]
```diff
@@ -3,13 +3,17 @@
...
+import asyncio
 
 import flask
 import tornado
 from tornado.httpserver import HTTPServer
 from tornado.ioloop import IOLoop
 from tornado.wsgi import WSGIContainer
 
+# asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())
+# needs to be uncommented for a Windows 10 + Python 3.8 config
```
Comment (bryevdv) [inline sphinx/docserver.py:15]: @nimishbongale I think we should just check platform and python version and do this automatically

**Comment by bryevdv** [comment]
@nimishbongale 

* The .rst source file for `index.html` is here: https://github.com/bokeh/bokeh/blob/master/sphinx/source/index.rst
* For `searchtools.js` I *think* if you add your own custom version here: https://github.com/bokeh/bokeh/tree/master/sphinx/source/bokeh/static it will get used. But I am not 100% sure
* The `search.html` is added by Sphinx during the build process. I am not sure how to override or modify it. Will have to dig in to the details  later this week

File: sphinx/docserver.py (hunk 1 above)
Comment (nimishbongale) [reply sphinx/docserver.py:15]: @bryevdv Okay, I agree.

Will make the necessary changes!

File: sphinx/docserver.py [hunk 2]
```diff
@@ -35,7 +41,7 @@ def send_alert():
 @app.route('/en/latest/<path:filename>')
 def send_docs(filename):
     return flask.send_from_directory(
-        os.path.join(_basedir, "sphinx/build/html/"), filename)
+        os.path.join(_basedir, "sphinx/_build/html/"), filename)
```
Comment (bryevdv) [inline sphinx/docserver.py:44]: @nimishbongale why this change? The subdir is named `build` on my local machine

File: sphinx/source/conf.py [hunk 3]
```diff
@@ -64,7 +64,7 @@
 
 autodoc_member_order = 'groupwise'
 
-bokeh_missing_google_api_key_ok = False
+bokeh_missing_google_api_key_ok = False #True
```
Comment (bryevdv) [inline sphinx/source/conf.py:67]: I am not sure this comment adds anything? Anyone who might need to change it will surely know that the value to use is True?

File: sphinx/source/index.rst.txt [hunk 4]
```diff
@@ -0,0 +1,45 @@
+.. _about:
```
Comment (bryevdv) [inline sphinx/source/index.rst.txt:1]: This file got saved with the wrong extension, it should be `index.rst` not `index.rst.txt`

File: sphinx/source/index.rst.txt [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+If you are interested in contributing to Bokeh, or extending the library,
+check out the :ref:`devguide` Guide.
+
+If you'd like to search for a particular topic, use the search box below:
+
+.. raw:: html
+
+    <form class="bd-search align-items-center" action="search.html" method="get">
+      <input type="search" class="form-control" name="q" id="search-input" placeholder="&#x1F50D; Search the docs ..." aria-label="Search the docs ..." autocomplete="off">
+    </form>
```
Comment (bryevdv) [inline sphinx/source/index.rst.txt:27]: BTW if you know how to make this search field not be so wide, I couldn't figure it out last time I tried :)

File: sphinx/source/bokeh/static/searchtools.js [hunk 6]
```diff
@@ -0,0 +1,534 @@
...
+    $.each(hlwords, function() {
+      rv = rv.highlightText(this, 'highlighted');
+    });
+    return rv;
+  }
+};
+
+$(document).ready(function() {
+  var para = document.getElementsByTagName('p')[1]
+  para.innerHTML+=`<br><input type="checkbox" id="concise">&nbsp;&nbsp;Check this option for faster, but shorter querying<br>`
```
Comment (bryevdv) [inline sphinx/source/bokeh/static/searchtools.js:523]: Would it be simple to just automatically show a page with "fast" results and then have a button to go to a page with "full" results? I think sometimes less choice is better, and I worry that this checkbox will be unfamiliar and confusing (or ignored) by users

File: sphinx/docserver.py (hunk 2 above)
Comment (nimishbongale) [reply sphinx/docserver.py:44]: @bryevdv it worked with `_build` in my machine. The folder that `sphinx-build` generated was named `_build`, and hence the change. Let me know if there's something wrong!

File: sphinx/docserver.py [hunk 7]
```diff
@@ -3,13 +3,19 @@
 import threading
 import time
 import webbrowser
+import asyncio
```
Comment (bryevdv) [inline sphinx/docserver.py:6]: import sort order (isort will complain)

File: sphinx/docserver.py (hunk 2 above)
Comment (bryevdv) [reply sphinx/docserver.py:44]: Yeah I checked and the Makefle specifies `BUILDDIR    = build` so I think this change needs to be reverted

File: sphinx/source/conf.py (hunk 3 above)
Comment (nimishbongale) [reply sphinx/source/conf.py:67]: @bryevdv before doing the final commit I just reverted back the changes, which I usually do by just commenting out the non-required boolean value. It can certainly be removed, I just kept it in thinking it doesn't do much harm!

File: sphinx/source/index.rst.txt (hunk 4 above)
Comment (nimishbongale) [reply sphinx/source/index.rst.txt:1]: Whops, my bad. Will change that.

File: sphinx/source/index.rst.txt (hunk 5 above)
Comment (nimishbongale) [reply sphinx/source/index.rst.txt:27]: @bryevdv can do it. Will make the necessary changes!

File: sphinx/source/bokeh/static/searchtools.js (hunk 6 above)
Comment (nimishbongale) [reply sphinx/source/bokeh/static/searchtools.js:523]: @bryevdv Hmm. Maybe. 

Introducing a button onto a page (which fills up quickly with search queries) I felt was a bad idea because if I position the button to appear after the search results, it makes no sense. The user will have to scroll to the end anyways. 

If I position the button next to the actual search button, users are bound to get confused which one to click on, and the UI doesn't look good when you have 2 buttons placed next to each other doing pretty much the same function. That's why I went with a checkbox, because that's the simplest way to get me a boolean value for my purposes. 

I hope the checkbox is not too badly placed/ ill-purposed. I tried multiple things to fit in better, but this was the one I liked the most.

File: sphinx/docserver.py (hunk 7 above)
Comment (nimishbongale) [reply sphinx/docserver.py:6]: @bryevdv 

Will fix! Wow, the number of checks surprised me, that's for sure!

File: sphinx/source/bokeh/static/searchtools.js (hunk 6 above)
Comment (bryevdv) [reply sphinx/source/bokeh/static/searchtools.js:523]: My main objection to the checkbox is that I can't think of any search context I've ever seen that has anything like that. It's very novel, and I think in this context novel is not desirable. If the fast results are limited to only show a smaller number (say less than 30) then I think scrolling to the bottom is not much of an issue? It seems very familiar/similar to me to e.g. scrolling to the bottom of Google results to get the to next page.

File: sphinx/source/conf.py (hunk 3 above)
Comment (bryevdv) [reply sphinx/source/conf.py:67]: Well, I wouldn't hold up merging if you really feel strongly about it, but my own preference is definitely to remove it.

File: sphinx/source/bokeh/static/searchtools.js (hunk 6 above)
Comment (bryevdv) [reply sphinx/source/bokeh/static/searchtools.js:523]: I should add: in my experience  the result I want when searching is usually in the top 10 or even 5 results, so I think limiting a "first page" to ~30 results would be more than enough to cover the vast majority of needs (and still make the "next page" button reasonable to get to for anyone that needs more)

Comment (nimishbongale) [reply sphinx/source/bokeh/static/searchtools.js:523]: @bryevdv You're right, in the sense that a checkbox isn't the most user friendly way to go about it. I wish I could do better than it, but I do believe it fixes the basic issue of slower searches, a slightly better UX. 

I think ultimately the bokeh docs search calls for good "pagination". At this time, what I've done is I've spliced the results and I'm showing the top half of them (for the fast query option). In case I've to show the top 30 results, I'd need to find a way to either:-

1. Hide the remaining results using some CSS and then upon a button click, reveal them.
2. Store the results somewhere and later append them into the page if the user wants more results to be shown.
3. Show the results page by page.

As of now, I spent some time getting familiar with sphinx and bokeh, and therefore made an attempt at improving the UX/Docs experince by a little bit. Yes, I did try the CSS approach too, but I couldn't get it to work. There must be some clever/smart way of doing it. I tried `nth-child` method so that all query results above 30 would instantly be hidden. It didn't work as per expectation. I should definitely look into how popular search engines do this kind of work so efficienty!

Option 3 would be absolutely great. Given time, I think its definitely worth a go. Maybe this could be  GSoD project idea in itself, coupled with something like "Accesibility a11y improvement". 

Thanks!

File: sphinx/source/conf.py (hunk 3 above)
Comment (nimishbongale) [reply sphinx/source/conf.py:67]: Alright, will remove it!

**Comment by nimishbongale** [comment]
@bryevdv, I've made all the necessary changes. Thank you so much for your guidance!
I'll be sure to work on the pagination logic soon, but it's bound to take some time. Will build my proposal for GSoD soon! Looking forward to it.

File: sphinx/docserver.py (hunk 2 above)
Comment (nimishbongale) [reply sphinx/docserver.py]: @bryevdv fixed it.

File: sphinx/source/conf.py (hunk 3 above)
Comment (nimishbongale) [reply sphinx/source/conf.py]: Removed it!

File: sphinx/source/index.rst.txt (hunk 4 above)
Comment (nimishbongale) [reply sphinx/source/index.rst.txt]: This stray file happened to appear our of nowhere! Removed it.

File: sphinx/source/index.rst.txt (hunk 5 above)
Comment (nimishbongale) [reply sphinx/source/index.rst.txt]: @bryevdv I changed the width, and centred the search bar. Looks much better now!

File: sphinx/docserver.py (hunk 7 above)
Comment (nimishbongale) [reply sphinx/docserver.py]: Fixed!

**Comment by bryevdv** [comment]
@nimishbongale I pushed a change to fix a linting issue but then LGTM came back with 3 alerts. I am not sure why they were not present before. Let me know if you can't access them on the LGTM link above. 

Also the unrelated build failures on master have just been fixed so they should be gone next push.

**Comment by nimishbongale** [comment]
> I am not sure why they were not present before.

@bryevdv This is because I added `searchtools.js` into the static folder. Some variables used in that are not defined within `searchtools.js`. I'm guessing they become available only once `sphinx-build` is run, and `searchtools.js` appears in build. I'm guessing LGTM just looks at the `static` copy of `searchtools.js` and gives it as a warning. 

This is what I've been able to understand. I am not 100% sure though.

**Comment by nitin10s** [comment]
My two cents on this would be that the checkbox is quite confusing. The prime reason for it being a bad user experience is the fact that search functionality is way too common that can be found nearly everywhere and associating something different with it will rather contribute to increasing the cognitive load of the users. This is in fact a well know UX law [[Read more](https://www.zorraquino.com/en/dictionary/ux/what-is-jakobs-law.html)] 

The descriptive text "Check this for faster, concise results" itself appears confusing.

**Comment by bryevdv** [comment]
@nimishbongale for the sake of separating concerns, can we table the search results work for now and just push the small changes to docserver, css, the placeholder,  etc for now? And then leave #10038 open to keep looking at more comprehensive solutions

**Comment by nimishbongale** [comment]
> My two cents on this would be that the checkbox is quite confusing. The prime reason for it being a bad user experience is the fact that search functionality is way too common that can be found nearly everywhere and associating something different with it will rather contribute to increasing the cognitive load of the users. This is in fact a well know UX law [[Read more](https://www.zorraquino.com/en/dictionary/ux/what-is-jakobs-law.html)]
> 
> The descriptive text "Check this for faster, concise results" itself appears confusing.

It will be fixed soon. Yes, it is an unorthodox way, and perhaps not the best way either. I didn't find it too hindering/cumbersome, but maybe that's just my opinion. From a user point of perspective, I can definitely see why users will find it confusing. I'll look into how I can find a workaround for this. 

Your suggestion is much appreciated!

**Comment by nimishbongale** [comment]
> @nimishbongale for the sake of separating concerns, can we table the search results work for now and just push the small changes to docserver, css, the placeholder, etc for now? And then leave #10038 open to keep looking at more comprehensive solutions

Yes. Will remove the checkbox associated codes until a better/clever way is found to tackle the issue.

**Comment by nimishbongale** [comment]
@bryevdv 

List of things done:

1. Added a magnifying lens icon to placeholder
2. Centred the search bar
3. Made the search bar smaller
4. Centred the gallery grid images
5. Changes to `docserver.py` to make it compatible (Win10 + Py3.8)

Things to do:
1. Efficient way to search and show results using a custom `searchtools.js` file.

**Comment by bryevdv** [comment]
I think it's just a flaky test, I will restart them to be safe

**Comment by bryevdv** [comment]
@mattpap what is going on with newly failing JS tests here? They were literally green when I restarted to re-run a flaky windows test and now they are suddenly failing on files that were not changed at all in this PR.

**Comment by mattpap** [comment]
This needs to be rebased on top of `master` (ref issue #10116).

**Comment by bryevdv** [comment]
> This needs to be rebased on top of master

I figured. For the sake of expediency, this seems safe enough to just merge as-is.

**Comment by bryevdv** [comment]
OK I fetched and ran codebase checks locally. They all pass so I will go ahead and merge.

**Comment by bryevdv** [comment]
Thanks, @nimishbongale

**Comment by nimishbongale** [comment]
> OK I fetched and ran codebase checks locally. They all pass so I will go ahead and merge.

Should it not be a problem though? I hope its okay! Was ready to rebase and merge.

**Comment by nimishbongale** [comment]
> > This needs to be rebased on top of master
> 
> I figured. ~For the sake of expediency, this seems safe enough to just merge as-is.~
> 
> Well it fails before the main codebase checks run, so I take that back. @nimishbongale can you merge/rebase on current master and we will get this merged?

I hope its all fine!

**Comment by nimishbongale** [comment]
> Thanks, @nimishbongale

Happy to contribute! Sorry for making this too long drawn out. Will work on the search optimization among other things soon!

Thanks for all the help @bryevdv! 👍

**Comment by bryevdv** [comment]
> Edit: Happened to notice the integration-tests failed. Anything about them?

We still have a few flaky tests, unfortunately. I don't think its related at all to your PR and  I've restarted it for good measure.

**Comment by nimishbongale** [comment]
Ahh okay alright!
